### PR TITLE
Session 3,4,5 PR

### PR DESCRIPTION
**Pull request by Mrinalaot**
Kindly review Session 3,4,5  Tasks

### CODE REVIEW

File: Session_3/closure.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//Closures and Scopes
+// http://www.codewars.com/kata/closures-and-scopes/train/javascript
+
+function createFunctions(n) {
+    var callbacks = [];
+
+    for (let i = 0; i < n; i++) {
```
Comment (AshisSarthak) [inline Session_3/closure.js]: Rename the variables. single word vars are very bad practice

File: Session_3/closure.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        callbacks.push(function() {
```
Comment (AshisSarthak) [inline Session_3/closure.js]: why not use es 6 ?

Just checking if es6 is done or not

File: Session_3/fill_array.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Prefill an Array
+// http://www.codewars.com/kata/prefill-an-array
+
+function prefill(n, v) {
+
+    if (isNaN(parseInt(n)) || n < 0 || n % 1 !== 0)
```
Comment (AshisSarthak) [inline Session_3/fill_array.js]: avoid multiple if else for returns

if(a === b){
return a;
}
return b:

is equivalent to 
if(a === b){
return a;
} else {
return b:
}

File: Session_3/fun_in_fun.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+//A function within a function
+// http://www.codewars.com/kata/a-function-within-a-function
+
+function always(func) {
```
Comment (AshisSarthak) [inline Session_3/fun_in_fun.js]: rewrite

const always = (func) => () => func

File: Session_3/private_set_get.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Can you keep a secret?
+// http://www.codewars.com/kata/can-you-keep-a-secret
+
+function createSecretHolder(secret) {
+    var _secret = secret;
```
Comment (AshisSarthak) [inline Session_3/private_set_get.js]: do not use var

File: Session_4/Cat_closure.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+// Let's make a Cat constructor!
+// http://www.codewars.com/kata/using-closures-to-share-class-state
+
+function Cat(name, weight) {
+    if (name == undefined || weight == undefined) {
```
Comment (AshisSarthak) [inline Session_4/Cat_closure.js]: this is a bad practice/
use if (!name || !weight)

File: Session_4/chain_add.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// A Chain adding function
+// http://www.codewars.com/kata/a-chain-adding-function
+
+function add(n1) {
+    var innerAdd = n2 => add(n1 + n2);
```
Comment (AshisSarthak) [inline Session_4/chain_add.js]: es6 var should never be used

File: Session_4/func_composition_2.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+function compose(...args) {
+
+    return function(a) {
+        let result = a;
+        for (let i = args.length - 1; i >= 0; i--) {
+            const fun = args[i];
+            result = fun(result)
+        }
+        return res;
```
Comment (AshisSarthak) [inline Session_4/func_composition_2.js]: where is res declared

File: Session_4/func_composition_2.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+// Function composition
+// http://www.codewars.com/kata/function-composition-1
+
+function compose(...args) {
+
+    return function(a) {
+        let result = a;
```
Comment (AshisSarthak) [inline Session_4/func_composition_2.js]: unnecessary usage of  result. use a directly or destructure it

File: Session_5/array_func.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// Array Helpers
+// http://www.codewars.com/kata/array-helpers
+
+Array.prototype.square = function() { return this.map(x => x * x) }
```
Comment (AshisSarthak) [inline Session_5/array_func.js]: you are combining es5 with es 6 

Array.prototype.square = function() { return this.map(x => x * x) } 

should be

Array.prototype.square = ()  => this.map(x => x * x)

**Review by AshisSarthak — CHANGES_REQUESTED** [review]

File: Session_3/closure.js (hunk 1 above)
Comment (Mrinalaot) [reply Session_3/closure.js]: In codewars default code content was like :
function createFunctions(n) {
  var callbacks = [];

  for (var i=0; i<n; i++) {
    callbacks.push(function() {
      return i;
    });
  }
  return callbacks;
}

Only I changed from var to let to show proper behavior as expected in test cases.

File: Session_3/closure.js (hunk 2 above)
Comment (Mrinalaot) [reply Session_3/closure.js]: es6 was not covered when I did this exercise. I will refactor this to es6

File: Session_5/array_func.js (hunk 10 above)
Comment (Mrinalaot) [reply Session_5/array_func.js]: By using "Array.prototype.square = () => this.map(x => x * x)" I was getting this error as "this" in arrow functions and normal function are not same. Please guide. 
TypeError: this.map is not a function
    at Array.square
    at /home/codewarrior/index.js:21:30

One query do we have to use arrow function forcefully in ES6, means can't mix old function definition?

**Review by Mrinalaot — COMMENTED** [review]
Modified as per (session 3,4,5) PR feedback. Please review once

File: Session_5/array_func.js (hunk 10 above)
Comment (AshisSarthak) [reply Session_5/array_func.js]: Arrow functions are mandated in es6+.

In this case try this.prototype.map/ here this refers to the current context. you can manipulate accordingly

**Review by AshisSarthak — APPROVED** [review]

**Pull request merged**